spec/core/v2/ics-004-packet-semantics/PACKET.md

```diff
@@ -0,0 +1,116 @@
+# Packet Specification
+
+## Packet V2
+
+The IBC packet sends application data from a source chain to a destination chain with a timeout that specifies when the packet is no longer valid. The packet will be committed to by the source chain as specified in the ICS-24 specification. The receiver chain will then verify the packet commitment under the ICS-24 specified packet commitment path. If the proof succeeds, the IBC handler sends the application data(s) to the relevant application(s).
+
+```typescript
+interface Packet {
+    // identifier on the source chain
+    // that source chain uses to address dest chain
+    // this functions as the sending address
+    sourceIdentifier: bytes,
+    // identifier on the dest chain
+    // that dest chain uses to address source chain
+    // this functions as the return address
+    destIdentifier: bytes,
+    // the sequence uniquely identifies this packet
+    // in the stream of packets from source to dest chain
+    sequence: uint64,
+    // the timeout is the timestamp in seconds on the destination chain
+    // at which point the packet is no longer valid.
+    // It cannot be received on the destination chain and can
+    // be timed out on the source chain
+    timeout: uint64,
+    // the data includes the messages that are intended
+    // to be sent to application(s) on the destination chain
+    // from application(s) on the source chain
+    // IBC core handlers will route the payload to the desired
+    // application using the port identifiers but the rest of the
+    // payload will be processed by the application
+    data: [Payload]
+}
+
+interface Payload {
+    // sourcePort identifies the sending application on the source chain
+    sourcePort: bytes,
+    // destPort identifies the receiving application on the dest chain
+    destPort: bytes,
+    // version identifies the version that sending application
+    // expects destination chain to use in processing the message
+    // if dest chain does not support the version, the payload must
+    // be rejected with an error acknowledgement
+    version: string,
+    // encoding allows the sending application to specify which
```

Comment:
On a more general note, I suggest we be very cautious about imposing `string` types at the protocol level. This can become costly in resource-limited environments like smart contracts. One field might not have a huge impact, though not being careful and leaving strings scattered throughout different data structures can increase memory allocations and overall costs due to operations like parsing.

Reply:
It helps a lot, but there is a cost to those as well.

To quickly illustrate with a dummy contract I wrote and tested in Remix:
```Solidity
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.20;

enum MimeType {
    JSON,
    PROTOBUF
}

contract GasTest {

    function stringEqualityCheck(string calldata encodingType) public pure returns (bool) {
        return keccak256(abi.encodePacked(encodingType)) == keccak256(abi.encodePacked("application/json"));
    }

    function enumEqualityCheck(MimeType encodingType) public pure returns (bool) {
        return encodingType == MimeType.JSON;
    }
}
```

The cost was 1327 and 593 for the string and enum check respectively.